tpm2/tpm2.go

```diff
@@ -0,0 +1,1247 @@
+	cmd, err := encodeIncrementNv(handle, authString)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, tagSessions, cmdIncrementNvCounter, tpmutil.RawBytes(cmd))
+	return err
+}
+
+func encodeUndefineSpace(owner, handle Handle) ([]byte, error) {
+	auth, err := encodePasswordAuthArea("", HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	out, err := tpmutil.Pack(owner, handle, uint16(0))
+	if err != nil {
+		return nil, err
+	}
+	out = append(out, auth...)
+	return out, nil
+}
+
+// NVUndefineSpace removes an index from TPM's NV storage.
+func NVUndefineSpace(rw io.ReadWriter, owner, handle Handle) error {
+	cmd, err := encodeUndefineSpace(owner, handle)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, tagSessions, cmdUndefineSpace, tpmutil.RawBytes(cmd))
+	return err
+}
+
+func encodeDefineSpace(owner, handle Handle, authString string, attributes uint32, policy []byte, dataSize uint16) ([]byte, error) {
+	pw, err := encodePasswordData(authString)
+	if err != nil {
+		return nil, err
+	}
+	auth, err := encodePasswordAuthArea("", HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	out1, err := tpmutil.Pack(owner, []byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	hashAlg := AlgSHA1
+	sizeNvArea := uint16(2*int(unsafe.Sizeof(owner)) + 3*int(unsafe.Sizeof(dataSize)) + len(policy))
+	out1 = append(append(out1, auth...), pw...)
+	out2, err := tpmutil.Pack(sizeNvArea, handle, hashAlg, attributes, policy, dataSize)
+	if err != nil {
+		return nil, err
+	}
+	return append(out1, out2...), nil
+}
+
+// NVDefineSpace creates an index in TPM's NV storage.
+func NVDefineSpace(rw io.ReadWriter, owner, handle Handle, authString string, policy []byte, attributes uint32, dataSize uint16) error {
+	cmd, err := encodeDefineSpace(owner, handle, authString, attributes, policy, dataSize)
+	if err != nil {
+		return err
+	}
+	_, err = runCommand(rw, tagSessions, cmdDefineSpace, tpmutil.RawBytes(cmd))
+	return err
+}
+
+func decodeNVReadPublic(in []byte) (NVPublic, error) {
+	var pub NVPublic
+	var buf []byte
+	if err := tpmutil.Unpack(in, &buf); err != nil {
+		return pub, err
+	}
+	err := tpmutil.Unpack(buf, &pub)
+	return pub, err
+}
+
+func decodeNVRead(in []byte) ([]byte, error) {
+	var sessionAttributes uint32
+	var data []byte
+	if err := tpmutil.Unpack(in, &sessionAttributes, &data); err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+func encodeNVRead(handle Handle, authString string, offset, dataSize uint16) ([]byte, error) {
+	out, err := tpmutil.Pack(handle, handle, []byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	auth, err := encodePasswordAuthArea(authString, HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	out = append(out, auth...)
+	params, err := tpmutil.Pack(dataSize, offset)
+	if err != nil {
+		return nil, err
+	}
+	out = append(out, params...)
+	return out, nil
+}
+
+// NVRead reads full data blob from NV index.
```

Comment:
grammar nit: "reads full data blob from NV index" -> "reads a full data blob from an NV index" (add the indefinite article twice)

Reply:
Done